messaging/src/main/java/org/axonframework/messaging/annotation/HierarchicalParameterResolverFactory.java

```diff
@@ -0,0 +1,84 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.messaging.annotation;
+
+import org.axonframework.common.Priority;
+
+import java.lang.reflect.Executable;
+import java.lang.reflect.Parameter;
+
+import static org.axonframework.common.Priority.LOW;
+
+/**
+ * {@link ParameterResolverFactory} that first tries to resolve a parameter using the child factory. If that fails, it
+ * tries the parent factory. This is useful to encapsulate a set of parameter resolvers that are only relevant in a
+ * specific context, such as a specific {@link org.axonframework.configuration.Module}.
+ *
+ * @author Mitchell Herrijgers
+ * @since 5.0.0
+ */
+@Priority(LOW)
+public class HierarchicalParameterResolverFactory implements ParameterResolverFactory {
+
+    private final ParameterResolverFactory parent;
+    private final ParameterResolverFactory child;
+
+    private HierarchicalParameterResolverFactory(ParameterResolverFactory parent, ParameterResolverFactory child) {
+        this.parent = parent;
+        this.child = child;
+    }
+
+    /**
+     * Creates a new hierarchical{@link ParameterResolverFactory} that delegates to the given {@code parent} and
+     * {@code child} factories. The {@code child} factory is tried first, and if it cannot resolve the parameter, the
+     * {@code parent} factory is tried.
+     *
+     * @param parent The parent {@link ParameterResolverFactory} to delegate to if the child cannot resolve the
+     *               parameter.
+     * @param child  The child {@link ParameterResolverFactory} to try first.
+     * @return A new hierarchical {@link ParameterResolverFactory} that delegates to the given factories.
+     */
+    public static HierarchicalParameterResolverFactory create(ParameterResolverFactory parent,
+                                                              ParameterResolverFactory child) {
+        return new HierarchicalParameterResolverFactory(parent, child);
+    }
+
+    @Override
+    public ParameterResolver<?> createInstance(Executable executable, Parameter[] parameters, int parameterIndex) {
+        ParameterResolver<?> resolver = child.createInstance(executable, parameters, parameterIndex);
+        if (resolver != null) {
+            return resolver;
+        }
+        return parent.createInstance(executable, parameters, parameterIndex);
+    }
+
+    /**
+     * Returns the parent {@link ParameterResolverFactory} of this factory.
+     * @return The parent {@link ParameterResolverFactory} of this factory.
+     */
+    public ParameterResolverFactory getParent() {
+        return parent;
+    }
+
+    /**
+     * Returns the child {@link ParameterResolverFactory} of this factory.
+     * @return The child {@link ParameterResolverFactory} of this factory.
+     */
+    public ParameterResolverFactory getChild() {
+        return child;
+    }
```

Comment:
You use those methods just in tests. So I think they are not needed at all (test on behavior - depends on which component will be returned with given parent/child), or you can make them package-private.